### PR TITLE
Move more classes into Internal namespaces.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- These classes aren't needed as top level public classes.
- There are several tests that don't follow our developer guidance; for these I'll take care of them as part of https://github.com/aspnet/Razor/issues/782.

#779

/cc @pranavkm

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor/Editor/Internal/BackgroundParser.cs [hunk 1]
```diff
@@ -9,13 +9,13 @@
...
+using Microsoft.AspNetCore.Razor.Parser.Internal;
 using Microsoft.AspNetCore.Razor.Parser.SyntaxTree;
 using Microsoft.AspNetCore.Razor.Text;
-using Microsoft.AspNetCore.Razor.Utils;
 
-namespace Microsoft.AspNetCore.Razor.Editor
+namespace Microsoft.AspNetCore.Razor.Editor.Internal
 {
-    internal class BackgroundParser : IDisposable
+    public class BackgroundParser : IDisposable
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor/Editor/Internal/BackgroundParser.cs]: cc @Eilon \ @rynowak for the `public` -> `internal` change.

File: src/Microsoft.AspNetCore.Razor/Parser/Internal/WhitespaceRewriter.cs [hunk 2]
```diff
@@ -24,36 +24,6 @@ protected override bool CanRewrite(Block block)
             return block.Type == BlockType.Expression && Parent != null;
         }
 
-        //public override void VisitBlock(Block block)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor/Parser/Internal/WhitespaceRewriter.cs]: :100:

File: test/Microsoft.AspNetCore.Razor.Test/Parser/RazorParserTest.cs [hunk 3]
```diff
@@ -12,6 +12,7 @@
 using Moq;
 using Moq.Protected;
 using Xunit;
+using Microsoft.AspNetCore.Razor.Parser.Internal;
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Razor.Test/Parser/RazorParserTest.cs]: Sort

File: test/Microsoft.AspNetCore.Razor.Test/Parser/RazorParserTest.cs [hunk 4]
```diff
@@ -74,8 +75,8 @@ public void GetTagHelperDescriptors_IsInvokedToLocateTagHelperDescriptors()
                                                Mock.Of<ITagHelperDescriptorResolver>());
             parser.CallBase = true;
             parser.Protected()
-                  .Setup<IEnumerable<TagHelperDescriptor>>("GetTagHelperDescriptors", 
-                                                           ItExpr.IsAny<Block>(), 
+                  .Setup<IEnumerable<TagHelperDescriptor>>("GetTagHelperDescriptors",
+                                                           ItExpr.IsAny<Block>(),
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Razor.Test/Parser/RazorParserTest.cs]: could you untab this while you're changing this code?

**Comment by pranavkm** [comment]
Changes look :shipit:. Just run the visibility change.by @Eilon \ @rynowak .

File: src/Microsoft.AspNetCore.Razor/Editor/AutoCompleteEditHandler.cs [hunk 5]
```diff
@@ -3,12 +3,13 @@
...
 using System.Collections.Generic;
-using Microsoft.AspNetCore.Razor.Editor;
+using Microsoft.AspNetCore.Razor.Parser;
+using Microsoft.AspNetCore.Razor.Parser.SyntaxTree;
 using Microsoft.AspNetCore.Razor.Text;
 using Microsoft.AspNetCore.Razor.Tokenizer.Symbols;
 using Microsoft.Extensions.Internal;
 
-namespace Microsoft.AspNetCore.Razor.Parser.SyntaxTree
+namespace Microsoft.AspNetCore.Razor.Editor
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor/Editor/AutoCompleteEditHandler.cs]: Why is this class moving?

File: src/Microsoft.AspNetCore.Razor/Editor/Internal/BackgroundParser.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor/Editor/Internal/BackgroundParser.cs]: @pranavkm this is just differently-`internal` now.

File: test/Microsoft.AspNetCore.Razor.Test/Utils/DisposableActionTest.cs [hunk 6]
```diff
@@ -2,7 +2,7 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 
-using Microsoft.AspNetCore.Razor.Utils;
+using Microsoft.AspNetCore.Razor.Parser.Internal;
 using Xunit;
 
 namespace Microsoft.AspNetCore.Razor.Test.Utils
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Razor.Test/Utils/DisposableActionTest.cs]: Should change the namespace of the test to match the tested class. No need for the added `using`.

File: test/Microsoft.AspNetCore.Razor.Test/Parser/WhitespaceRewriterTest.cs [hunk 7]
```diff
@@ -2,6 +2,7 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using Microsoft.AspNetCore.Razor.Parser;
+using Microsoft.AspNetCore.Razor.Parser.Internal;
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Razor.Test/Parser/WhitespaceRewriterTest.cs]: Fix the `namespace` instead of adding this `using`.

File: test/Microsoft.AspNetCore.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs [hunk 8]
```diff
@@ -15,6 +15,7 @@
 using Xunit;
 using Microsoft.AspNetCore.Razor.Parser.TagHelpers.Internal;
 using Microsoft.AspNetCore.Razor.Parser.TagHelpers;
+using Microsoft.AspNetCore.Razor.Parser.Internal;
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs]: Fix the `namespace` instead of adding this `using`.

File: test/Microsoft.AspNetCore.Razor.Test/Text/TextReaderExtensionsTest.cs [hunk 9]
```diff
@@ -4,6 +4,7 @@
 using System;
 using System.IO;
 using Microsoft.AspNetCore.Razor.Parser;
+using Microsoft.AspNetCore.Razor.Parser.Internal;
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Razor.Test/Text/TextReaderExtensionsTest.cs]: Change the `namespace` instead of adding this `using`.

File: test/Microsoft.AspNetCore.Razor.Test/TagHelpers/TagHelperDirectiveSpanVisitorTest.cs [hunk 10]
```diff
@@ -6,9 +6,9 @@
 using System.Linq;
 using Microsoft.AspNetCore.Razor.Parser;
 using Microsoft.AspNetCore.Razor.Parser.SyntaxTree;
-using Microsoft.AspNetCore.Razor.Parser.TagHelpers;
+using Microsoft.AspNetCore.Razor.Parser.TagHelpers.Internal;
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Razor.Test/TagHelpers/TagHelperDirectiveSpanVisitorTest.cs]: Fix the `namespace` instead of adding this `using`.

File: src/Microsoft.AspNetCore.Razor/Text/Internal/LineTrackingStringBuffer.cs [hunk 11]
```diff
@@ -7,9 +7,9 @@
 using System.Text;
 using Microsoft.AspNetCore.Razor.Parser;
 
-namespace Microsoft.AspNetCore.Razor.Text
+namespace Microsoft.AspNetCore.Razor.Text.Internal
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor/Text/Internal/LineTrackingStringBuffer.cs]: Curious: Other than changes to match classes you're moving, why change any test `namespace`s?

File: src/Microsoft.AspNetCore.Razor/Tokenizer/Internal/CSharpKeywordDetector.cs [hunk 12]
```diff
@@ -5,7 +5,7 @@
 using System.Collections.Generic;
 using Microsoft.AspNetCore.Razor.Tokenizer.Symbols;
 
-namespace Microsoft.AspNetCore.Razor.Tokenizer
+namespace Microsoft.AspNetCore.Razor.Tokenizer.Internal
 {
     internal static class CSharpKeywordDetector
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor/Tokenizer/Internal/CSharpKeywordDetector.cs]: `public` if you're changing the `namespace`.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNetCore.Razor/Editor/AutoCompleteEditHandler.cs (hunk 5 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor/Editor/AutoCompleteEditHandler.cs]: It never belonged in the SyntaxTree namespace. It's an Editor feature, that's where the rest of the edithandlers are.

File: test/Microsoft.AspNetCore.Razor.Test/Utils/DisposableActionTest.cs (hunk 6 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Test/Utils/DisposableActionTest.cs]: See description, this is done in a different PR. Same for other comments like this one.

File: src/Microsoft.AspNetCore.Razor/Text/Internal/LineTrackingStringBuffer.cs (hunk 11 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor/Text/Internal/LineTrackingStringBuffer.cs]: To match that of the corresponding src its testing.

File: src/Microsoft.AspNetCore.Razor/Tokenizer/Internal/CSharpKeywordDetector.cs (hunk 12 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor/Tokenizer/Internal/CSharpKeywordDetector.cs]: Will unchange namespace, talked to Eilon and might as well just keep the internal guys internal and where they wer eoriginally

File: src/Microsoft.AspNetCore.Razor/Editor/Internal/BackgroundParser.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor/Editor/Internal/BackgroundParser.cs]: Talked with Eilon. Should keep this `internal class`

**Comment by NTaylorMullen** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Razor/Text/LineTrackingStringBuffer.cs [hunk 13]
```diff
@@ -114,7 +114,7 @@ private TextLine ScanLines(int absoluteIndex, int startPos)
             return null;
         }
 
-        internal struct CharacterReference
+        public struct CharacterReference
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor/Text/LineTrackingStringBuffer.cs]: Separate file?

File: src/Microsoft.AspNetCore.Razor/Tokenizer/Internal/CSharpKeywordDetector.cs (hunk 12 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor/Tokenizer/Internal/CSharpKeywordDetector.cs]: > might as well just keep the internal guys internal and where they were originally [typo corrected]

I'm confused by the `namespace` change here if the decision was to leave `internal` classes where they were. Also, does that decision cover `src` as well as `test`. A few classes in both have moved and become `public`. This and `DisposableAction` moved but aren't `public`.

File: src/Microsoft.AspNetCore.Razor/Text/LineTrackingStringBuffer.cs (hunk 13 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor/Text/LineTrackingStringBuffer.cs]: Or should this remain `internal`?

File: src/Microsoft.AspNetCore.Razor/Tokenizer/Internal/CSharpKeywordDetector.cs (hunk 12 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor/Tokenizer/Internal/CSharpKeywordDetector.cs]: Something's up with my box. The namespace should have been nuked. Running through this all locally again.

**Comment by NTaylorMullen** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Razor/Editor/RazorEditorTrace.cs [hunk 14]
```diff
@@ -9,7 +9,7 @@
 
 namespace Microsoft.AspNetCore.Razor.Editor
 {
-    internal static class RazorEditorTrace
+    public static class RazorEditorTrace
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor/Editor/RazorEditorTrace.cs]: So close

**Comment by dougbu** [comment]
:shipit: after restoring `RazorEditorTrace`'s `internal`ity.
